pkg/controller/controller_broker.go

```diff
@@ -336,14 +341,29 @@ func (c *controller) reconcileClusterServiceBroker(broker *v1alpha1.ClusterServi
 			return stderrors.New(s)
 		}
 
-		for _, servicePlan := range servicePlans {
-			glog.V(4).Infof("ClusterServiceBroker %q: reconciling ClusterServicePlan (K8S: %q ExternalName: %q)", broker.Name, servicePlan.Name, servicePlan.Spec.ExternalName)
-			if err := c.reconcileClusterServicePlanFromClusterServiceBrokerCatalog(broker, servicePlan); err != nil {
+		// get the existing services and plans for this broker so that we can
+		// detect when services and plans are removed from the broker's
+		// catalog
+		existingServiceClasses, existingServicePlans, err := c.getCurrentServiceClassesAndPlansForBroker(broker)
+		if err != nil {
+			return err
+		}
+
+		existingServiceClassMap := convertServiceClassListToMap(existingServiceClasses)
+		existingServicePlanMap := convertServicePlanListToMap(existingServicePlans)
+
+		// reconcile the plans that were part of the broker's catalog payload
+		for _, payloadServicePlan := range payloadServicePlans {
+			existingServicePlan, _ := existingServicePlanMap[payloadServicePlan.Name]
```

Comment:
This makes me think how do we handle the situation when `service` was renamed on the broker's side, but the `plan` remained the same. My guess is:
- plan will be marked as healthy
- but class will be marked with `RemovedFromBrokerCatalog = true`

Will this situation be handled correctly by controllers?

Reply:
Remember - if the plan IDs are the same - they will be updated.